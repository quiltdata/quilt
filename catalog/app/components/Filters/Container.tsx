import cx from 'classnames'
import * as React from 'react'
import * as M from '@material-ui/core'
import { fade } from '@material-ui/core/styles'

import Skeleton from 'components/Skeleton'

const useStyles = M.makeStyles((t) => ({
  root: {},
  close: {},
  content: {
    display: 'flex',
    flexDirection: 'column',
    padding: t.spacing(0, 0, 2),
    position: 'relative',
  },
  header: {
    alignItems: 'center',
    display: 'flex',
    justifyContent: 'space-between',
    margin: t.spacing(0, 0, 1.5),
  },
  title: {
    ...t.typography.body2,
    fontWeight: 500,
  },
  lock: {
    alignItems: 'center',
    animation: '$showLock .3s ease-out',
    background: fade(t.palette.background.paper, 0.7),
    bottom: 0,
    display: 'flex',
    justifyContent: 'center',
    left: 0,
    position: 'absolute',
    right: 0,
    top: 0,
  },
  spinner: {},
  '@keyframes showLock': {
    '0%': {
      transform: 'scale(1.2x)',
    },
    '100%': {
      transform: 'scale(1)',
    },
  },
}))

interface ContainerProps {
  children?: React.ReactNode
  className?: string
  extenting?: boolean
  onDeactivate?: () => void
  title: React.ReactNode
  defaultExpanded?: boolean
}

export default function Container({
  className,
  children,
  title,
  extenting,
  onDeactivate,
}: ContainerProps) {
  const classes = useStyles()
  return (
    <div className={cx(classes.root, className)}>
      <div className={classes.header}>
        <div className={classes.title}>{title}</div>
        {onDeactivate && (
          <M.IconButton size="small" className={classes.close} onClick={onDeactivate}>
            <M.Icon fontSize="inherit">clear</M.Icon>
          </M.IconButton>
        )}
      </div>
      <div className={classes.content}>
        {children ? (
          <>
            {children}
            {extenting && (
              <div className={classes.lock}>
                <M.CircularProgress className={classes.spinner} size={28} />
              </div>
            )}
          </>
        ) : (
          <Skeleton height={32} />
        )}
      </div>
    </div>
  )
}
