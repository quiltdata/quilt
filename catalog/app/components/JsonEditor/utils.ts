import { JsonValue } from './constants'

export const stringifyJSON = (obj: JsonValue) => JSON.stringify(obj, null, 2)

export function parseJSON(str: string) {
  try {
    return JSON.parse(str)
  } catch (e) {
    return str
  }
}
