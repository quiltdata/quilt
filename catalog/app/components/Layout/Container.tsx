import cx from 'classnames'
import invariant from 'invariant'
import * as React from 'react'
import * as M from '@material-ui/core'

const Ctx = React.createContext<{
  state: boolean
  adjustCounter: (inc: number) => void
} | null>(null)

interface FullWidthProviderProps {
  children: React.ReactNode
}

export function FullWidthProvider({ children }: FullWidthProviderProps) {
  const [state, setState] = React.useState(false)
  const counterRef = React.useRef(0)
  const adjustCounter = React.useCallback((inc: number) => {
    counterRef.current += inc
    setState(counterRef.current > 0)
  }, [])
  return <Ctx.Provider value={{ state, adjustCounter }}>{children}</Ctx.Provider>
}

export function useSetFullWidth() {
  const ctx = React.useContext(Ctx)
  invariant(ctx, 'Context must be used within a Provider')
  const { adjustCounter } = ctx
  React.useEffect(() => {
    adjustCounter(1)
    return () => {
      adjustCounter(-1)
    }
  }, [adjustCounter])
}

export function useGetFullWidth() {
  const ctx = React.useContext(Ctx)
  invariant(ctx, 'Context must be used within a Provider')
  return ctx.state
}

const useStyles = M.makeStyles((t) => ({
  fullWidth: {
    animation: t.transitions.create('$expand'),
  },
  contained: {
    animation: t.transitions.create('$collapse'),
  },
  '@keyframes expand': {
    '0%': {
      transform: 'scaleX(0.94)',
    },
    '100%': {
      transform: 'scaleX(1)',
    },
  },
  '@keyframes collapse': {
    '0%': {
      opacity: 0.3,
    },
    '100%': {
      opacity: 1,
    },
  },
}))

export function Container({ children, className }: M.ContainerProps) {
  const classes = useStyles()

  const fullWidth = useGetFullWidth()

  return (
    <M.Container
      className={cx(fullWidth ? classes.fullWidth : classes.contained, className)}
      maxWidth={!fullWidth && 'lg'}
    >
      {children}
    </M.Container>
  )
}

export default Container
