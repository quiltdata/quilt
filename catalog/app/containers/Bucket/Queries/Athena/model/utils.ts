export const Loading = Symbol('loading')

export type Maybe<T> = T | null

// `T` is the value
// `undefined` is no data. It is not initialized
// `Loading` is loading
// `Error` is error
export type Data<T> = T | undefined | typeof Loading | Error

export interface DataController<T> {
  data: Data<T>
  loadMore: () => void
}

export function wrapData<T>(data: Data<T>, setPrev: (d: T) => void): DataController<T> {
  return {
    data,
    loadMore: () => hasData(data) && setPrev(data),
  }
}

export interface List<T> {
  list: T[]
  next?: string
}

// `T` is the value
// `null` is no value, explicitly set by user
// `undefined` is no value. It is not initialized
// `Loading` is loading
// `Error` is error
export type Value<T> = Maybe<Data<T>>

export interface ValueController<T> {
  value: Value<T>
  setValue: (v: T | null) => void
}

export function wrapValue<T>(
  value: Value<T>,
  setValue: (d: T | null) => void,
): ValueController<T> {
  return {
    value,
    setValue,
  }
}

/** Data is loaded, or value is set to actual value */
export function hasData<T>(value: Value<T>): value is T {
  if (
    value === undefined ||
    value === Loading ||
    value instanceof Error ||
    value === null
  ) {
    return false
  }
  return true
}

/** No value yet: value or data was just initialized */
export function isNone<T>(value: Value<T>): value is undefined {
  return value === undefined
}

/** Data is loading, or value is waiting for data */
export function isLoading<T>(value: Value<T>): value is typeof Loading {
  return value === Loading
}

export function isError<T>(value: Value<T>): value is Error {
  return value instanceof Error
}

/** Value is selected with some or no value, or resolved with error, or data is loaded (successfully or not) */
export function isReady<T>(value: Value<T>): value is T | null | Error {
  if (value === undefined || value === Loading) {
    return false
  }
  return true
}

/** Value is selected with some or no value, or data is loaded successfully */
export function hasValue<T>(value: Value<T>): value is T | null {
  if (value === undefined || value === Loading || value instanceof Error) {
    return false
  }
  return true
}

/** User explicitly set no value */
export function isNoneSelected<T>(value: Value<T>): value is null {
  return value === null
}
