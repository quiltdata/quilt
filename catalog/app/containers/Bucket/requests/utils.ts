import * as R from 'ramda'

export const decodeS3Key = R.pipe(R.replace(/\+/g, ' '), decodeURIComponent)
