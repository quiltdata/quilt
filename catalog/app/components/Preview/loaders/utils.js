import { extname } from 'path'

import * as R from 'ramda'
import * as React from 'react'

import * as AWS from 'utils/AWS'
import AsyncResult from 'utils/AsyncResult'
import * as Config from 'utils/Config'
import Data from 'utils/Data'
import * as NamedRoutes from 'utils/NamedRoutes'

import { PreviewError } from '../types'

export const SIZE_THRESHOLDS = [
  128 * 1024, // automatically load if <= 128kB
  1024 * 1024, // never load if > 1MB
]

export const COMPRESSION_TYPES = { gz: '.gz', bz2: '.bz2' }

// eslint-disable-next-line consistent-return
export const getCompression = (key) => {
  // eslint-disable-next-line no-restricted-syntax
  for (const [type, ext] of Object.entries(COMPRESSION_TYPES)) {
    if (key.endsWith(ext)) return type
  }
}

export const stripCompression = (key) => {
  const comp = getCompression(key)
  return comp ? key.slice(0, -COMPRESSION_TYPES[comp].length) : key
}

export const extIs = (ext) => (key) => extname(key).toLowerCase() === ext

export const extIn = (exts) => (key) => exts.includes(extname(key).toLowerCase())

export const withSigner = (callback) => <AWS.Signer.Inject>{callback}</AWS.Signer.Inject>

export const withS3Request = (callback) => (
  <AWS.S3.InjectRequest>{callback}</AWS.S3.InjectRequest>
)

export const withRoutes = (callback) => (
  <NamedRoutes.Inject>{callback}</NamedRoutes.Inject>
)

export const withData = (props, callback) => <Data {...props}>{callback}</Data>

const gate = async ({ s3req, handle }) => {
  let length
  try {
    const head = await s3req({
      bucket: handle.bucket,
      operation: 'headObject',
      params: {
        Bucket: handle.bucket,
        Key: handle.key,
        VersionId: handle.version,
      },
    })
    length = head.ContentLength
  } catch (e) {
    if (['NoSuchKey', 'NotFound'].includes(e.name)) {
      throw PreviewError.DoesNotExist({ handle })
    }
    // eslint-disable-next-line no-console
    console.error('Error loading preview')
    // eslint-disable-next-line no-console
    console.error(e)
    throw PreviewError.Unexpected({ handle, originalError: e })
  }
  if (length > SIZE_THRESHOLDS[1]) {
    throw PreviewError.TooLarge({ handle })
  }
  return { handle, gated: length > SIZE_THRESHOLDS[0] }
}

export const gatedS3Request = (fetcher) => (handle, callback, extraParams) =>
  withS3Request((s3req) =>
    withData(
      { fetch: gate, params: { s3req, handle } },
      AsyncResult.case({
        Ok: ({ gated }) =>
          fetcher({ s3req, handle, gated, ...extraParams }, (r, ...args) =>
            callback(AsyncResult.Ok(r), ...args),
          ),
        _: callback,
      }),
    ),
  )

const parseRange = (range) => {
  if (!range) return undefined
  const m = range.match(/bytes \d+-\d+\/(\d+)$/)
  if (!m) return undefined
  return Number(m[1])
}

const getFirstBytes = (bytes) => async ({ s3req, handle }) => {
  try {
    const res = await s3req({
      bucket: handle.bucket,
      operation: 'getObject',
      params: {
        Bucket: handle.bucket,
        Key: handle.key,
        VersionId: handle.version,
        Range: `bytes=0-${bytes}`,
      },
    })
    const firstBytes = res.Body.toString('utf-8')
    const contentLength = parseRange(res.ContentRange) || 0
    return { firstBytes, contentLength }
  } catch (e) {
    if (['NoSuchKey', 'NotFound'].includes(e.name)) {
      throw PreviewError.DoesNotExist({ handle })
    }
    // eslint-disable-next-line no-console
    console.error('Error loading preview')
    // eslint-disable-next-line no-console
    console.error(e)
    throw PreviewError.Unexpected({ handle, originalError: e })
  }
}

export const withFirstBytes = (bytes, fetcher) => {
  const fetch = getFirstBytes(bytes)

  return (handle, callback) =>
    withS3Request((s3req) =>
      withData(
        { fetch, params: { s3req, handle } },
        AsyncResult.case({
          Ok: ({ firstBytes, contentLength }) =>
            fetcher({ s3req, handle, firstBytes, contentLength }, (r, ...args) =>
              callback(AsyncResult.Ok(r), ...args),
            ),
          _: callback,
        }),
      ),
    )
}

export const objectGetter = (process) => {
  const fetch = ({ s3req, handle, ...extra }) =>
    s3req({
      bucket: handle.bucket,
      operation: 'getObject',
      params: {
        Bucket: handle.bucket,
        Key: handle.key,
        VersionId: handle.version,
      },
    }).then((r) => process(r, { s3req, handle, ...extra }))

  return ({ s3req, handle, gated, ...extra }, callback) =>
    withData({ fetch, params: { s3req, handle, ...extra }, noAutoFetch: gated }, callback)
}

const previewUrl = (endpoint, query) =>
  `${endpoint}/preview${NamedRoutes.mkSearch(query)}`

const fetchPreview = async ({ endpoint, type, handle, signer, ...rest }) => {
  const signed = signer.getSignedS3URL(handle)
  const compression = getCompression(handle.key)
  const r = await fetch(
    previewUrl(endpoint, { url: signed, input: type, compression, ...rest }),
  )
  const json = await r.json()
  return json
}

const withGatewayEndpoint = (callback) => (
  <Config.Inject>
    {AsyncResult.case({
      Err: R.pipe(
        AsyncResult.Err,
        callback,
      ),
      _: callback,
      Ok: R.pipe(
        R.prop('apiGatewayEndpoint'),
        AsyncResult.Ok,
        callback,
      ),
    })}
  </Config.Inject>
)

export const previewFetcher = (type, process) => {
  const fetch = (x) => fetchPreview(x).then((res) => process(res, x))
  return (handle, callback, extra) =>
    withSigner((signer) =>
      withGatewayEndpoint(
        AsyncResult.case({
          Err: (e, ...args) => {
            const pe = PreviewError.Unexpected({ handle, originalError: e })
            return callback(AsyncResult.Err(pe), ...args)
          },
          Ok: (endpoint) =>
            withData(
              { fetch, params: { endpoint, type, handle, signer, ...extra } },
              AsyncResult.case({
                _: callback,
                Err: (e, ...args) => {
                  const pe = PreviewError.Unexpected({ handle, originalError: e })
                  return callback(AsyncResult.Err(pe), ...args)
                },
              }),
            ),
          _: callback,
        }),
      ),
    )
}
