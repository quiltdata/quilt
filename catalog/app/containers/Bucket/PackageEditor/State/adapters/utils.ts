import type { FileWithPath } from 'react-dropzone'

import { Status } from 'components/FileManager/FileRow'
import type { TreeEntry } from 'components/FileManager/FileTree'
import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

export function calcChildren(
  entry: Model.PackageEntry | Model.S3File | FileWithPath,
  tailParts: string[],
  children: TreeEntry[] = [],
): TreeEntry[] {
  const [name, ...tail] = tailParts
  const found = children.find((child) => child.id === name)
  if (found) {
    if (tail.length) {
      found.children = calcChildren(entry, tail, found.children)
    }
    return children
  }

  return children.concat({
    id: name,
    name: tail.length ? s3paths.ensureSlash(name) : name,
    size: tail.length ? 0 : entry.size,
    status: Status.Unchanged,
    children: tail.length ? calcChildren(entry, tail, []) : undefined,
  })
}

export function sortEntries(entries: TreeEntry[]): TreeEntry[] {
  return [...entries].sort((a, b) => {
    if (a.children && !b.children) return -1
    if (!a.children && b.children) return 1
    return a.name.localeCompare(b.name)
  })
}
