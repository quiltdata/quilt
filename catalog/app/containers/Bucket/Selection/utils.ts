import { join } from 'path'

import * as R from 'ramda'

import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

export interface SelectionItem {
  logicalKey: string
}

export interface ListingSelection {
  [prefixUrl: string]: SelectionItem[]
}

export const EMPTY_MAP: ListingSelection = {}

interface SelectionHandles {
  [prefixUrl: string]: Model.S3.S3ObjectLocation[]
}

const convertIdToHandle = (
  id: string | number,
  parentHandle: Model.S3.S3ObjectLocation,
): Model.S3.S3ObjectLocation => ({
  bucket: parentHandle.bucket,
  key: join(parentHandle.key, id.toString()),
})

export const toHandlesMap = (selection: ListingSelection): SelectionHandles =>
  Object.entries(selection).reduce(
    (memo, [prefixUrl, items]) => ({
      ...memo,
      [prefixUrl]: items.map((item) =>
        convertIdToHandle(item.logicalKey, s3paths.parseS3Url(prefixUrl)),
      ),
    }),
    {} as SelectionHandles,
  )

export const toHandlesList = (selection: ListingSelection): Model.S3.S3ObjectLocation[] =>
  Object.entries(selection).reduce(
    (memo, [prefixUrl, items]) => [
      ...memo,
      ...items.map((item) =>
        convertIdToHandle(item.logicalKey, s3paths.parseS3Url(prefixUrl)),
      ),
    ],
    [] as Model.S3.S3ObjectLocation[],
  )

const mergeWithFiltered =
  (prefix: string, filteredItems: SelectionItem[]) => (allItems: SelectionItem[]) => {
    if (!allItems || !allItems.length) return filteredItems
    const selectionOutsideFilter = allItems.filter(
      (item) => !item.logicalKey.startsWith(prefix),
    )
    const newIds = [...selectionOutsideFilter, ...filteredItems]
    return R.equals(newIds, allItems) ? allItems : newIds // avoids cyclic update
  }

export function merge(
  items: SelectionItem[],
  bucket: string,
  path: string,
  filter?: string,
): (state: ListingSelection) => ListingSelection {
  const prefixUrl = `s3://${bucket}/${path}`
  const lens = R.lensProp<Record<string, SelectionItem[]>>(prefixUrl)
  return filter ? R.over(lens, mergeWithFiltered(filter, items)) : R.set(lens, items)
}

const EmptyKeys: string[] = []

export const getDirectorySelection = (
  selection: ListingSelection,
  bucket: string,
  path: string,
) => selection[`s3://${bucket}/${path}`] || EmptyKeys
