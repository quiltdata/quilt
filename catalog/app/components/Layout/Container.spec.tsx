import * as React from 'react'
import renderer from 'react-test-renderer'

import { createBoundary } from 'utils/ErrorBoundary'

import { FullWidthProvider, Container, useSetFullWidth } from './Container'

const ErrorBoundary = createBoundary(() => (error: Error) => (
  <span>Error: {error.message}</span>
))

const EmptyContainer = () => (
  <ErrorBoundary>
    <Container>{''}</Container>
  </ErrorBoundary>
)

describe('components/Layout/Container', () => {
  it('requires Provider', () => {
    jest.spyOn(console, 'error').mockImplementationOnce(jest.fn())
    const tree = renderer.create(<EmptyContainer />)
    expect(tree).toMatchSnapshot()
  })

  it('has restricted width by default', () => {
    const tree = renderer.create(
      <FullWidthProvider>
        <EmptyContainer />
      </FullWidthProvider>,
    )
    expect(tree.root.findByProps({ maxWidth: 'lg' })).toBeTruthy()
    expect(() => tree.root.findByProps({ maxWidth: false })).toThrow()
    expect(tree).toMatchSnapshot()
  })

  it('has full width once set', () => {
    const SetFullWidth = () => {
      useSetFullWidth()
      return <>long content</>
    }
    const tree = renderer.create(
      <FullWidthProvider>
        <Container>
          <SetFullWidth />
        </Container>
      </FullWidthProvider>,
    )
    renderer.act(() => {})
    expect(() => tree.root.findByProps({ maxWidth: 'lg' })).toThrow()
    expect(tree.root.findByProps({ maxWidth: false })).toBeTruthy()
    expect(tree).toMatchSnapshot()
  })

  it('still has full width when other remove full width', () => {
    // TODO: Show some warning or throw error:
    //       Some component is designed for narrow layout,
    //       but is rendered in `fullWidth`,
    //       because there are rendered other components designed for `fullWidth`.

    const SetFullWidth = () => {
      useSetFullWidth()
      return <>long content</>
    }
    const UnmountSetFullWidth = () => {
      const [x, setX] = React.useState(true)
      React.useEffect(() => {
        setX(false)
      }, [])
      return x ? <SetFullWidth /> : <>short content</>
    }
    const tree = renderer.create(
      <FullWidthProvider>
        <Container>
          <UnmountSetFullWidth />
          <SetFullWidth />
          <UnmountSetFullWidth />
          <SetFullWidth />
          <UnmountSetFullWidth />
        </Container>
      </FullWidthProvider>,
    )
    renderer.act(() => {})
    expect(() => tree.root.findByProps({ maxWidth: 'lg' })).toThrow()
    expect(tree.root.findByProps({ maxWidth: false })).toBeTruthy()
    expect(tree).toMatchSnapshot()
  })
})
