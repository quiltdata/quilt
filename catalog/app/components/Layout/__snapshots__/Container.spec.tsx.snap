// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Layout/Container has full width once set 1`] = `
<div
  class="MuiContainer-root makeStyles-fullWidth-1"
>
  long content
</div>
`;

exports[`components/Layout/Container has restricted width by default 1`] = `
<div
  class="MuiContainer-root makeStyles-contained-2 MuiContainer-maxWidthLg"
/>
`;

exports[`components/Layout/Container requires Provider 1`] = `
<span>
  Error: 
  Context must be used within a Provider
</span>
`;

exports[`components/Layout/Container still has full width when other remove full width 1`] = `
<div
  class="MuiContainer-root makeStyles-fullWidth-1"
>
  short content
  long content
  short content
  long content
  short content
</div>
`;
